### PR TITLE
fix(output_parsers): iterable unpacking compatibility in content safety parsers

### LINKED ISSUE #1241: bug: investigate parser malformed output causing >2-tuple returns
### Did you check docs and existing issues?

- [x] I have read all the NeMo-Guardrails docs
- [x] I have updated the package to the latest version before submitting this issue
- [x] (optional) I have used the develop branch
- [x] I have searched the existing issues of NeMo-Guardrails

### Python version (python --version)

Python 3.12

### Operating system/version

MacOS 15.3.2 (24D81)

### NeMo-Guardrails version (if you must use a specific version and not the latest

0.14.0

## Problem Statement
While fixing the starred unpacking compatibility in #1207 , we discovered that content safety parsers are sometimes returning malformed data structures that don't match their documented interface.

## Root Cause Analysis Needed
The recent fix addressed the symptom (starred unpacking crashes) but not the underlying issue: **parsers returning >2-tuples instead of proper 2-tuples**.

### Expected vs Actual Behavior

**✅ Expected (works fine):**
```python
(True, [])                    # safe content
(False, ['s3', 's22', 's23']) # unsafe with violations
```

**❌ Actual (breaks old code):**
```python
(False, 's3', 's22', 's23')  # 4-tuple instead of 2-tuple
```

## Impact
- **Immediate**: Fixed by PR #<PR-number> with Sequence return type
- **Long-term**: Inconsistent data structures flowing through the system
- **Risk**: Other parts of codebase may expect 2-tuple format

### Steps To Reproduce

see #1207 

### Expected Behavior

should always return a sequence with first element being boolean so that iterable unpacking works

The content safety actions use iterable unpacking syntax now:
This syntax expects the parser to return a sequence where:
* Safe content: [True] → unpacks to is_safe=True, violated_policies=[]
* Unsafe content: [False, "S1", "S8"] → unpacks to is_safe=False, violated_policies=["S1", "S8"]

### Actual Behavior

```python
# Original parsers returned tuples:
(True, [])                    # Safe content
(False, ['S1', 'S8'])        # Unsafe with violations  
(False, [''])                # Bug: empty violations as [""] not []

```

### PR DESCRIPTION
**Pull request by Pouyanpi**
## Problem
Content safety actions were failing due to output parser interface incompatibility. The parsers returned `Tuple[bool, List[str]]` format, but actions used starred unpacking syntax `is_safe, *violated_policies = result` which expected list format.

## Root Cause
Multiple bugs in output parsers:
1. **Iterable Unpacking Bug**: Tuple format `(True, [])` incompatible with `is_safe, *violated_policies = result` that is introduced in #1207 
2. **Empty Violations Bug**: Returned `[""]` instead of `[]` for no violations
3. **Case Sensitivity Bug**: Violations converted to lowercase, losing original case (`"s1"` vs `"S1"`)

## Solution
-  **Changed return type**: `Tuple[bool, List[str]]` → `Sequence[Union[bool, str]]`
-  **Fixed starred unpacking**: Now returns `[True]` or `[False, "policy1", "policy2"]` format
-  **Preserved violation case**: `"S1"` stays `"S1"`, not `"s1"`
-  **Added helper function**: `_parse_unsafe_violations()` for robust parsing
-  **Tests**: tests covering edge cases and different scenarios
-

## Impact
- Content safety actions now work correctly with iterable unpacking syntax
- Better violation parsing with preserved case information  
- Type system compliance (resolved covariance issues)
- Complete test coverage for future maintainability

## Breaking Changes
None: this is a bug fix that makes the newly introduced interable unpacking syntax work as intended.

Closes #1241

### CODE REVIEW

File: tests/test_content_safety_integration.py [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+        context = {"user_message": "Dubious violent content"}
+
+        result = await content_safety_check_input(
+            llms=llms,
+            llm_task_manager=mock_task_manager,
+            model_name="test_model",
+            context=context,
+        )
+
+        # TODO: @trebedea @prasoonvarshney is this the expected behavior?
```
Comment (Pouyanpi) [inline tests/test_content_safety_integration.py:158]: please have a look

File: tests/test_content_safety_integration.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+        context = _create_input_context("Dubious violent content")
+
+        result = await content_safety_check_input(
+            llms=llms,
+            llm_task_manager=mock_task_manager,
+            model_name="test_model",
+            context=context,
+        )
+
+        # TODO: @trebedea @prasoonvarshney is this the expected behavior?
```
Comment (Pouyanpi) [inline tests/test_content_safety_integration.py]: @trebedea @prasoonvarshney

**Review by cparisien — APPROVED** [review]
This looks ok to me, I'd still like to see @trebedea or @prasoonvarshney check on the questions from Pouyan.

**Comment by trebedea** [comment]
Actually, this is a newly introduced bug when fixing this one:
https://github.com/NVIDIA/NeMo-Guardrails/pull/1207 

The first fix was fine, then somebody suggested using starred unpacking syntax and the final MR opted for this. We should have had the tests from the beginning to catch these bugs. GG for adding the tests @Pouyanpi !

File: nemoguardrails/llm/output_parsers.py [hunk 3]
```diff
@@ -98,33 +112,33 @@ def is_content_safe(response: str) -> Tuple[bool, List[str]]:
         response (str): The response string to analyze.
 
     Returns:
-        Tuple[bool, Optional[List[str]]]: A tuple where the first element is a boolean indicating the safety of the content (True if safe, False otherwise),
-        and the second element is a list of violated policies, if any.
+        List[Union[bool, str]]: A list where the first element is a boolean indicating the safety of the content (True if safe, False otherwise),
```
Comment (trebedea) [inline nemoguardrails/llm/output_parsers.py]: Is there a reason why in the docstring we opt for `List` and in the type hints for `Sequence`?

**Comment by Pouyanpi** [comment]
@trebedea thanks for the review. Yes,you are right, #1207 was a tactical fix which was not addressing the root cause, all these subsequent PRs try to ensure that content safety is more robust 🤞🏻

File: nemoguardrails/llm/output_parsers.py (hunk 3 above)
Comment (trebedea) [reply nemoguardrails/llm/output_parsers.py]: Maybe we should also clean the doctring in this refactoring - I feel at least L104-107 are from an older version.

File: tests/test_content_safety_integration.py (hunk 2 above)
Comment (trebedea) [reply tests/test_content_safety_integration.py]: Looks good!

File: nemoguardrails/llm/output_parsers.py (hunk 3 above)
Comment (Pouyanpi) [reply nemoguardrails/llm/output_parsers.py]: Yes I forgot it 👍🏻  Thank you

File: nemoguardrails/llm/output_parsers.py [hunk 4]
```diff
@@ -176,8 +194,8 @@ def nemoguard_parse_response_safety(response: str) -> Tuple[bool, List[str]]:
...
```
Comment (Pouyanpi) [inline nemoguardrails/llm/output_parsers.py]:
```suggestion
        Sequence[Union[bool, str]]: A sequence where the first element is a boolean indicating the safety of the content (True if safe, False otherwise),
```

File: nemoguardrails/llm/output_parsers.py [hunk 5]
```diff
@@ -138,8 +152,8 @@ def nemoguard_parse_prompt_safety(response: str) -> Tuple[bool, List[str]]:
...
```
Comment (Pouyanpi) [inline nemoguardrails/llm/output_parsers.py]:
```suggestion
        Sequence[Union[bool, str]]: A sequence where the first element is a boolean indicating the safety of the content (True if safe, False otherwise),
```

File: nemoguardrails/llm/output_parsers.py (hunk 3 above)
Comment (Pouyanpi) [inline nemoguardrails/llm/output_parsers.py]:
```suggestion
        Sequence[Union[bool, str]]: A sequence where the first element is a boolean indicating the safety of the content (True if safe, False otherwise),
```

File: tests/test_content_safety_integration.py (hunk 2 above)
Comment (Pouyanpi) [inline tests/test_content_safety_integration.py]:
```suggestion
```
